flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -196,6 +197,17 @@ public void dispose() throws Exception {
 	public void close() throws Exception {
 		super.close();
 
+		waitReaderFinishing();
+
+		output.close();
+	}
+
+	@Override
+	public void endInput() throws Exception {
+		waitReaderFinishing();
```

Comment:
I think the same fix is needed in `AsyncWaitOperator`

Reply:
I don't like the feel that fixing similar issues one by one :(
Let's discuss under JIRA ticket.